File: Daily/2024-04/2024-04-19/17299.오등큰수.py
import sys; input = sys.stdin.readline
n = int(input())
arr = list(map(int, input().split()))
num = {}
for i in range(n):
    num[arr[i]] = num.get(arr[i], 0) + 1
q = []
ans = [-1] * n
for i in range(n):
    print(q)
    if not q:
        q.append((i, num[arr[i]]))
    else:
        while q:
            idx, cnt = q.pop()
            if cnt < num[arr[i]]:
                ans[idx] = arr[i]
            else:
                q.append((idx, cnt))
                break
        q.append((i, num[arr[i]]))
print(*ans)

File: Daily/2024-04/2024-04-19/17299.오등큰수.rs
use std::{cmp, io, vec};
use std::io::prelude::*;

fn read<T>(si: &mut T) -> String where T: Read {
    let mut s = String::new();
    si.read_to_string(&mut s).unwrap();
    s
}

fn next<T>(it: &mut std::str::SplitAsciiWhitespace) -> T where
    T: std::str::FromStr,
    <T as std::str::FromStr>::Err: std::fmt::Debug {
    it.next().unwrap().parse().unwrap()
}


fn main() {
    // 입력
    let mut si = io::BufReader::new(io::stdin().lock());
    let mut so = io::BufWriter::new(io::stdout().lock());
    let s = read(&mut si);
    let mut it = s.split_ascii_whitespace();

    let n = next::<usize>(&mut it);
    let mut arr = (0..n).map(|_| next::<usize>(&mut it)).collect::<Vec<_>>();
    let mut q = Vec::<(usize,usize)>::new();
    let mut ans = (0..n).map(|_| -1).collect::<Vec<_>>();

    let mut num = (0..*arr.iter().max().unwrap() as usize + 1).map(|_| 0).collect::<Vec<_>>();
    for i in 0..n {
        num[arr[i]] += 1;
    }
    for i in 0..n {
        if q.is_empty() {
            q.push((i, num[arr[i]]));
        } else {
            while !q.is_empty() {
                let (idx, cnt) = q[q.len()-1];
                if cnt < num[arr[i]] {
                    ans[idx] = arr[i] as isize;
                    q.pop();
                } else {
                    break;
                }
            }
            q.push((i, num[arr[i]]));
        }
    }
    for i in 0..n {
        write!(so, "{} ", ans[i]).ok();
    }
}